phabricatoremails/render/template.py

```diff
@@ -0,0 +1,290 @@
+        text = self._text_template.render(**template_params)
+
+        return self._css_inline.transform(html, False), text
+
+
+class TemplateStore:
+    """Configures Jinja and exposes the templates
+
+    There's two sets of templates: text and html. Each set has its own
+    distinct configuration and set of custom tests and filters.
+    """
+
+    def __init__(
+        self, css_text: str, keep_css_classes: bool, html_loader=None, text_loader=None
+    ):
+        self._css_inline = Premailer(
+            css_text=css_text,
+            exclude_pseudoclasses=False,
+            align_floating_images=False,
+            disable_leftover_css=False,
+            disable_validation=True,
+            remove_classes=not keep_css_classes,
+            allow_network=False,
+            strip_important=False,
+        )
+        self.html_jinja_env = _jinja_html(
+            html_loader
+            if html_loader
+            else jinja2.PackageLoader("phabricatoremails", "render/templates/html")
+        )
+        self.text_jinja_env = _jinja_text(
+            text_loader
+            if text_loader
+            else jinja2.PackageLoader("phabricatoremails", "render/templates/text")
+        )
+
+    def get(self, template_name: str) -> Template:
+        """Returns the Jinja template at the specific relative path within
+        the "templates" directory
+        """
+
+        return Template(
+            self._css_inline,
+            self.html_jinja_env.get_template(f"{template_name}.html.jinja2"),
+            self.text_jinja_env.get_template(f"{template_name}.text.jinja2"),
+        )
+
+
+def _jinja_html(loader):
+    jinja_env = jinja2.Environment(
+        loader=loader,
+        autoescape=True,
+        # If a template uses a variable that's not defined, throw an error
+        undefined=jinja2.StrictUndefined,
+        # the templates won't be changed at runtime, don't watch for changes
+        auto_reload=False,
+        # Cache all templates (rather than cycling the number of cached ones)
+        cache_size=-1,
+    )
+    jinja_env.tests["reply"] = _is_reply
+    jinja_env.tests["code"] = _is_code
+    jinja_env.tests["accepted"] = _is_reviewer_accepted
+    jinja_env.tests["blocking"] = _is_reviewer_blocking
+    jinja_env.tests["soft_needed"] = _is_reviewer_soft_needed
+    jinja_env.tests["needed"] = _is_reviewer_needed
+    jinja_env.tests["commented_on"] = _is_commented_on
+    jinja_env.tests["commented_on_secure"] = _is_commented_on_secure
+    jinja_env.filters["date"] = _date
+    jinja_env.filters["diff_class"] = _diff_class
+    jinja_env.filters["diff_symbol"] = _diff_symbol
+    jinja_env.filters["reviewer_actionable_class"] = _reviewer_actionable_class
+    jinja_env.filters["reviewer_status_icon"] = _reviewer_status_icon
+    jinja_env.filters["existence_change_class"] = _existence_change
+    jinja_env.filters["existence_change_label"] = _existence_change
```

Comment:
Are these intentionally the same value? If so, why?

Reply:
Yes, because they have very similar logic, but mean two different things.

From the template's perspective: it wants to set a CSS class according to the "existence" state of a reviewer. Later, it wants to provide a text label for this reviewer.

From the implementation perspective, the CSS classes currently conveniently line up with the text label, capitalization and all. I put them into a shared function because you mentioned duplication earlier. However, they should appear as two different filters to the templates.